icicle-compiler/test/Icicle/Test/Sea/Zebra.hs

```diff
@@ -0,0 +1,449 @@
+zebraOfWellTyped wt =
+  case zebraOfFacts (wtFactType wt) (fmap atFact (wtFacts wt)) of
+    Left e ->
+      Savage.error (show e)
+    Right Nothing ->
+      return Nothing
+    Right (Just (ty, tombstones, table)) -> do
+      -- FIXME ignoring fact times for now, but to test it we should convert icicle time to 1600 epoch secs here
+      -- let ts = fmap (Zebra.Time . fromIntegral . Icicle.secondsCountJulian . atTime) (wtFacts wt)
+      let ts = List.replicate (length (wtFacts wt)) 0
+      ps <- vectorOf (length ts) Zebra.jFactsetId
+      let attribute = Zebra.Attribute (Storable.fromList ts) (Storable.fromList ps) (Storable.fromList tombstones) table
+      entity <- uncurry Zebra.Entity <$> Zebra.jEntityHashId <*> pure (Boxed.singleton attribute)
+      pure . Just $ ZebraWellTyped wt ty entity
+
+-- we are not reading nested arrays in zebra right now
+schemaOfType' :: ValType -> Maybe Schema
+schemaOfType' ty = case ty of
+  ArrayT {} ->
+    Nothing
+  BufT {} ->
+    Nothing
+  MapT {} ->
+    Nothing
+  _ ->
+    schemaOfType ty
+
+schemaOfType :: ValType -> Maybe Schema
+schemaOfType ty = case ty of
+  BoolT ->
+    pure Schema.Bool
+
+  TimeT ->
+    pure Schema.Int
+
+  DoubleT ->
+    pure Schema.Double
+
+  IntT ->
+    pure Schema.Int
+
+  StringT ->
+    pure $ Schema.Array Schema.Byte
+
+  ErrorT ->
+    pure Schema.Int
+
+  UnitT ->
+    Nothing
+
+  FactIdentifierT ->
+    pure Schema.Int
+
+  ArrayT t ->
+    Schema.Array <$> schemaOfType' t
+
+  BufT _ t ->
+    Schema.Array <$> schemaOfType' t
+
+  PairT a b -> do
+    a' <- schemaOfType a
+    b' <- schemaOfType b
+    pure . Schema.Struct . Boxed.fromList $
+      [ Schema.Field (Schema.FieldName "fst") a'
+      , Schema.Field (Schema.FieldName "snd") b' ]
+
+  OptionT t ->
+    let
+      noneOf x =
+        Schema.Variant (Schema.VariantName "none") <$> schemaOfType x
+      someOf x =
+        Schema.Variant (Schema.VariantName "some") <$> schemaOfType x
+    in Schema.Enum <$> noneOf t <*> (Boxed.singleton <$> someOf t)
+
+  SumT a b ->
+    let
+      leftOf x =
+        Schema.Variant (Schema.VariantName "left")  <$> schemaOfType x
+      rightOf x =
+        Schema.Variant (Schema.VariantName "right") <$> schemaOfType x
+    in Schema.Enum <$> leftOf a <*> (Boxed.singleton <$> rightOf b)
+
+  MapT k v -> do
+    k' <- schemaOfType' k
+    v' <- schemaOfType' v
+    pure . Schema.Struct . Boxed.fromList $
+      [ Schema.Field (Schema.FieldName "keys") (Schema.Array k')
+      , Schema.Field (Schema.FieldName "vals") (Schema.Array v') ]
+
+  StructT struct
+    -- An empty icicle struct is not unit (unlike zebra), it has no input variable.
+    | fields <- getStructType struct
+    , not (Map.null fields) ->
+        let
+          fieldOf (f, t) =
+            Schema.Field (Schema.FieldName (nameOfStructField f)) <$> schemaOfType t
+        in Schema.Struct . Boxed.fromList <$> mapM fieldOf (Map.toList fields)
+
+  _ -> Nothing
```

Comment:
I think we should be able to map all of `ValType` to zebra? probably doesn't matter for a test I guess